auth/couchdb_session_authenticator_test.go

```diff
@@ -0,0 +1,281 @@
+
+		auth, err = NewCouchDbSessionAuthenticatorFromMap(map[string]string{
+			"USERNAME": username,
+			"PASSWORD": password,
+		})
+		Expect(err).To(BeNil())
+		Expect(auth).ToNot(BeNil())
+		Expect(auth.AuthenticationType()).To(Equal(AUTHTYPE_COUCHDB_SESSION))
+	})
+
+	It("Test Authenticator instantiation failures", func() {
+		errortests := []struct {
+			user, password string
+		}{
+			{"", "password"},
+			{"{invalid-user}", "password"},
+			{"user", ""},
+			{"user", "{invalid-password}"},
+		}
+
+		for _, tt := range errortests {
+			_, err := NewCouchDbSessionAuthenticator(tt.user, tt.password)
+			Expect(err).To(HaveOccurred())
+
+			_, err = NewCouchDbSessionAuthenticatorFromMap(map[string]string{
+				"USERNAME": tt.user,
+				"PASSWORD": tt.password,
+			})
+			Expect(err).To(HaveOccurred())
+		}
+	})
+
+	Context("with standard test server", func() {
+		var (
+			server  *httptest.Server
+			request *http.Request
+			auth    *CouchDbSessionAuthenticator
+			err     error
+		)
+
+		BeforeEach(func() {
+			callNumber := 0
+			server = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				callNumber++
+				cookie := &http.Cookie{
+					Name:  "AuthSession",
+					Value: fmt.Sprintf("fakefake-%d", callNumber),
+					Expires: time.Now().
+						Add(24*time.Hour - time.Duration(callNumber)*time.Minute),
+				}
+				http.SetCookie(w, cookie)
+				w.WriteHeader(http.StatusOK)
+			}))
+
+			builder, err := core.NewRequestBuilder(core.GET).
+				ResolveRequestURL(server.URL, "/db", nil)
+			Expect(err).To(BeNil())
+
+			ctx := context.WithValue(context.Background(), contextKey("key"), "abcdefgh")
+			request, err = builder.
+				AddHeader("X-Req-Id", "abcdefgh").
+				WithContext(ctx).
+				Build()
+			Expect(err).To(BeNil())
+			Expect(request).ToNot(BeNil())
+
+			auth, err = NewCouchDbSessionAuthenticator("user", "pass")
+			Expect(err).To(BeNil())
+			Expect(auth).ToNot(BeNil())
+			Expect(auth.AuthenticationType()).To(Equal(AUTHTYPE_COUCHDB_SESSION))
+		})
+
+		AfterEach(func() {
+			server.Close()
+		})
+
+		It("Test setting URL, Headers and Context on Authenticator", func() {
+			err = auth.Authenticate(request)
+			Expect(err).To(BeNil())
+
+			Expect(auth.url).To(Equal(server.URL))
+			Expect(auth.header).To(HaveKeyWithValue("X-Req-Id", []string{"abcdefgh"}))
+			Expect(auth.ctx.Value(contextKey("key"))).To(Equal("abcdefgh"))
+		})
+
+		It("Test setting custom http client on Authenticator", func() {
+			// set http.Client with custom timeout
+			auth.Client = &http.Client{Timeout: time.Second}
+
+			err = auth.Authenticate(request)
+			Expect(err).To(BeNil())
+
+			Expect(auth.Client.Timeout).To(Equal(time.Second))
+		})
+
+		It("Test authentication re-request after expiration", func(done Done) {
+			err = auth.Authenticate(request)
+			Expect(err).To(BeNil())
+
+			// Fetch a cookie from the cache to confirm it's present
+			// and prepare cache state for the next test
```

Comment:
What is preparation for the cache state?
```suggestion
			// Fetch a cookie from the session cache to confirm it's present
```

Reply:
Cache state includes cookie, its expiration time and its refresh time. In this test we are not as much interested in presence of the cookie, but in times around its management.